File: Programmers/JiHo/1주차(Greedy)/큰 수 만들기.py
#시간복잡도 O(n^2)
def solution(number, k):
    k = len(number) - k # K수만큼제거 -> K만큼 수를 선택
    answer = ''
    max_index, search_index = 0, 0
    while k > 0 : #K만큼 수를 선택
        max = 0
        for i in range(search_index, len(number) - k + 1) : #시작점(가장큰수 다음)~끝점(선택할수있는범위)까지 탐색
            if max < int(number[i]) : #큰값을 찾으면
                max, max_index = int(number[i]), i #큰값의 인덱스, 값 저장
                if max == 9 : break #테스트케이스 8,10 해결
        search_index = max_index + 1 #다음 시작점 설정
        answer += str(max)
        k -= 1 
    return answer

File: Programmers/Segun/1주차(Greedy)/큰 수 만들기.cpp
//(0부터 k+1까지)인덱스 0 부터 k+1 까지 중에서 가장 큰수와 그 값의 인덱스를 저장한다
//가장 큰수를 찾은 그 인덱스 전까지 범위를 k에서 빼준다
// k의 값과 탐색지점을 기억해주고 다시 반복한다(max값의 인덱스 부터 k+1까지) 

#include <string>
#include <vector>
#include<algorithm>

using namespace std;

string solution(string number, int k) {
    string answer = "";
    int i=0;  //0부터 시작
    int size=number.size()-k; //찾아야하는 문자열의 크기
  
    while(1){
        char mx='0';   //문자 비교이기때문에 char 선언
        int idx=i;  //idx 초기화
      
        for(int j=i;j<=i+k;j++){  //k의 범위만큼 탐색
            if(mx<number[j]){mx=number[j];idx=j;} //max값과 그 값의 인덱스 저장
        }
        answer+=number[idx];  //answer에 max값 추가
        if(answer.size()==size){return answer;} //answer가 다 채워졌다면 return
        
        k-=(idx-i); //k 재설정
        i=idx+1;  //i 재설정
    }
}
